### PR TITLE
File operation events support multiple resources

### LINKED ISSUE #98309: Support multiple files in working copy file service
When dragging multiple files in the explorer, the `onWillRenameFiles` event is fired multiple times concurrently with a single file per-call - even though the API supports an array:

https://github.com/microsoft/vscode/blob/926fc23f139e4d1250dba494c6e825e4816e62f6/src/vs/workbench/api/common/extHostFileSystemEventService.ts#L185

This means an extension will produce multiple edits for a single operation if there are multiple files moved - though it looks like VS Code merges them together:

https://github.com/microsoft/vscode/blob/926fc23f139e4d1250dba494c6e825e4816e62f6/src/vs/workbench/api/common/extHostFileSystemEventService.ts#L219-L228

However, this fails if two of the edits modified the same file with "(file) has changed in the meantime".

Here's a gif showing individual file moves correctly updating the file on the right. However if I moved them both at the same time, only one edit is applied and the other one fails.

multi_rename_fails2

The code for that repro is available here:

https://github.com/DanTup/vscode-repro-will-rename-files ([extension.ts](https://github.com/DanTup/vscode-repro-will-rename-files/blob/b9c41b2c954c371efed83860c93e75dbe25586ce/src/extension.ts)).

### PR DESCRIPTION
**Pull request by pfongkye**
This PR fixes #98309 

Hi @bpasero 
I made some changes. Can you tell me if I'm on the right track?
If I am, I will continue with the following:

- [x] add support move/copy multiple resources in workingCopyFileService (wcfs)
- [x] add support delete multiple resources in wcfs
- [x] support for onWillRunFileOperation in wcfs
- [x] support for onDidRunFileOperation in wcfs?
- [x] adoption in explorerViewer
- [x] adoption in fileActions?
~~- [ ] adoption in bulkEditService?~~
- [x] add tests if any
- [x] come up with a better name than `UriComponentsPair`?
- [x] check if it fixes given [repro](https://github.com/DanTup/vscode-repro-will-rename-files)

Let me know if there is anything missing on the above list.
Thanks.

### CODE REVIEW

**Comment by bpasero** [comment]
@jrieken for this code change that touches the extension side of things.

**Comment by pfongkye** [comment]
Indeed, I'll make the necessary changes in working copy file service also. Thanks for the feedback.

**Comment by bpasero** [comment]
One quick feedback that I noticed, I think it is fine to have just one method in the working copy file service that accepts an array instead of having a `move` and `moveMany` method.

**Comment by pfongkye** [comment]
Yes. I was thinking the same thing.

**Comment by pfongkye** [comment]
I think the PR is ok for review. Let me know if we should support [onDidRunFileOperation](https://github.com/microsoft/vscode/blob/master/src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts#L270) in this PR or should I create another issue?

**Comment by pfongkye** [comment]
Btw I think we can support the same operations for [delete ](https://github.com/microsoft/vscode/pull/98988/files#diff-59b5411e06f5ae92cd9e7a46c25a16bbL275) also. What do you think?

For the [bulkEdit](https://github.com/microsoft/vscode/pull/98988/files#diff-dd28630510373b817bce7b052a0b7a96L336), I'm not sure if `options.overwrite` will be the same for all the resources.

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts
Comment (bpasero) [inline src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:1306]: Prefer `for (const source of sources)`

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts
Comment (bpasero) [inline src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:1321]: Prefer `for (const source of sources)`

**Review by bpasero — CHANGES_REQUESTED** [review]

**Comment by bpasero** [comment]
@pfongkye I had a quick look and think you are on the right path. Once you think this PR is ready for review, I would add maybe also Isidor to review the changes for where this needs to be adopted (file explorer land).

Yes, I think for consistency I would also do this for `delete` 👍

**Comment by bpasero** [comment]
Assigning some folks to review:
* @jrieken for changes in ext host and bulk edit
* @isidorn for the adoption in explorer land (file actions, explorer)
* me for the working copy changes

File: src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts:51]: Can we extract this into a reusable type since we seem to be using that also above:

```ts
readonly files: {
   readonly source?: URI;
   readonly target: URI;
}[]
```

File: src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts:221]: This cannot be untyped.

File: src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts:226]: Missing `await` on `canMove`

File: src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts:235]: Type information missing

File: src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts:244]: Missing await

File: src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts:246]: Missing await

**Review by bpasero — CHANGES_REQUESTED** [review]
I feel that maybe the file operations inside `workingCopyFileService` should run in sequence and not in parallel to reduce the chance of overlapping file system operations? I think that would also make the code cleaner.

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts
Comment (isidorn) [inline src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:1326]: Please type this array

Comment (isidorn) [reply src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:1326]: Also can we use a more explicit name, `files` sounds too general

**Review by isidorn — CHANGES_REQUESTED** [review]

**Comment by isidorn** [comment]
Great PR, thanks for working on this!
I did an initial review of the Explorer land and the changes look good overall. Once you have tackled mine and @bpasero comments I will test it out and give it one more review. Thanks

**Review by jrieken — COMMENTED** [review]
👍 from me

**Comment by pfongkye** [comment]
Thank you all for the feedback. I'll make the necessary changes :+1:

File: src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts
Comment (pfongkye) [reply src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts]: I couldn't find a better name than `SourceTargetPair` type.

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts
Comment (pfongkye) [reply src/vs/workbench/contrib/files/browser/views/explorerViewer.ts]: The array is typed and used a more explicit name, I think.

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts
Comment (pfongkye) [inline src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:1374]: Should we have finer control on the files to be overwritten? For now, the message warns that all files/folders being moved will be overwritten.

Comment (bpasero) [reply src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:1374]: @isidorn

**Comment by bpasero** [comment]
Converting back to draft until you think this is ready for review again 👍

**Comment by pfongkye** [comment]
> I feel that maybe the file operations inside `workingCopyFileService` should run in sequence and not in parallel to reduce the chance of overlapping file system operations? I think that would also make the code cleaner.

I think that with the missing `await`s added in `delete` and `moveOrCopy`, the instructions seem to run sequentially. Or am I missing something else? @bpasero

**Comment by bpasero** [comment]
@pfongkye yes, when you `await` the file operations, they will.

**Comment by pfongkye** [comment]
> Converting back to draft until you think this is ready for review again 👍

I think the PR will be ready for review as soon as I get clarifications from @isidorn on what should be the approach for [this](https://github.com/microsoft/vscode/pull/98988/files#r440491841).

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts
Comment (isidorn) [reply src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:1384]: Idealy the dialog would list all the files which would be overwritten and there would just be a single confirm for all of them.
Having a fine grainined choice may end up being a lot of dialogs, so for now I would prefer the single confirm approach but at least let the user know exactly what files are being overwritten.

**Comment by isidorn** [comment]
I appologise, I somehow missed that question. Now I replied. Thanks for the ping.

**Comment by pfongkye** [comment]
> I appologise, I somehow missed that question. Now I replied. Thanks for the ping.

No worries. I'll make the necessary changes 👍

**Comment by pfongkye** [comment]
I think it's ok for review now.

**Review by bpasero — APPROVED** [review]
Pushed a few :lipstick: changes (via https://github.com/microsoft/vscode/pull/98988/commits/2338c817c81f75cfa43366af3b85b2a4cc7365a4), otherwise looks good to me, but @isidorn wanted to have a final review of the explorer related changes.

File: src/vs/workbench/contrib/files/browser/fileActions.ts
Comment (isidorn) [inline src/vs/workbench/contrib/files/browser/fileActions.ts:1036]: Changes in this file look great, good job.

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts
Comment (isidorn) [inline src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:726]: It's weird that details is passed into getMultipleFilesOverwriteConfirm and then concatenated, why not pass in the files and build the message properly

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts
Comment (isidorn) [inline src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:1360]: Maybe doHandleExplorerDrop would be easier to understand if it would split between code for copy operation vs move operation.

Comment (isidorn) [reply src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:1360]: This is an optional change which I can tackle also once we merge this PR

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts
Comment (isidorn) [inline src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:1369]: You open editors one by one when copied, you should use the openEditors method to bulk open imho

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts
Comment (isidorn) [inline src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:1399]: Maybe the helper method gets all the overwrites and then just creates an appropriate confirm.
So you do not need the length > 1 logic outside the method

**Review by isidorn — CHANGES_REQUESTED** [review]

**Comment by isidorn** [comment]
@pfongkye thanks for your work.
@bpasero and me reviewed it and we have provided the comments together. Once you tackle them we plan to merge.
Also note that I am going on vacation soon and would like to merge this in hopefully tomorrow :)

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts
Comment (pfongkye) [reply src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:726]: I'll do the changes 👍

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts
Comment (pfongkye) [reply src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:1360]: I'll see what I can do before you go on vacation :)

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts
Comment (pfongkye) [reply src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:1369]: I'll look into this also :+1:

**Comment by pfongkye** [comment]
> @pfongkye thanks for your work.
> @bpasero and me reviewed it and we have provided the comments together. Once you tackle them we plan to merge.
> Also note that I am going on vacation soon and would like to merge this in hopefully tomorrow :)
Thanks again for the feedback. I'll try to tackle the necessary changes hopefully today :+1:

**Comment by isidorn** [comment]
Great, thanks a lot! Let me know when it is ready for review so we merge this in 🚀

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts
Comment (pfongkye) [reply src/vs/workbench/contrib/files/browser/views/explorerViewer.ts]: I've split into two separate functions.

**Comment by pfongkye** [comment]
I've taken into account your comments. I think it's good for another review :+1:

**Review by isidorn — APPROVED** [review]

**Comment by bpasero** [comment]
Thanks 👍

**Pull request merged**
